add action to link users to accounts

nora-codecov opened the pull request:
### Purpose/Motivation
Gives us a method to do the initial setup when an Account is created. This action will create the link between the Account object and the User object for any Owner who is a `plan_activated_user` on any of the Organizations linked to the Account.

Can be run multiple times safely (in case it errors or times out)
Won't do the link if the Account doesn't have enough seats

For "auditing" its behavior - it gives messages about how many Users and AccountUsers were created. You can see the Users linked to the Account on the Account edit page (when you click into an Account on the admin). You can remove a AccountUsers connection on the User admin page.

### Links to relevant tickets
Uses https://github.com/codecov/shared/pull/295
part of https://github.com/codecov/engineering-team/issues/2058

adrian-codecov commented on codecov_auth/admin.py:
```diff
@@ -310,6 +311,66 @@ class AccountAdmin(AdminMixin, admin.ModelAdmin):
+    @admin.action(description="Link Users to Account")
+    def link_users_to_account(self, request, queryset):
+        for account in queryset:
+            all_plan_activated_user_ids = set()
```
Nit/Optional: I find the "all" a bit redundant since the variable is plural. I also got tripped up a bit reading this like "(all plan) (activated user ids)" instead  of "(all) (plan activated user ids)" 😅, but that might be me. I think calling this "activated_user_ids_in_orgs" represents this a bit better? 

Similar idea for the rest of the "alls" as well

adrian-codecov commented on codecov_auth/admin.py:
```diff
@@ -310,6 +311,66 @@ class AccountAdmin(AdminMixin, admin.ModelAdmin):
+                ownerid__in=all_plan_activated_user_ids
+            ).prefetch_related("user")
+            non_student_count = all_plan_activated_users.exclude(student=True).count()
+            total_seats_for_account = account.plan_seat_count + account.free_seat_count
```
What's a free seat count? I see it's defaulted to 0, is this for like free plans?

nora-codecov replied:
it's a thing that exists on Owners currently (named `free`), but it's a way for us to give Orgs additional seats not in the count they pay for (`plan_seat_count`) 🤷

adrian-codecov replied:
damn, was not aware of this

adrian-codecov commented on codecov_auth/admin.py:
```diff
@@ -310,6 +311,66 @@ class AccountAdmin(AdminMixin, admin.ModelAdmin):
+            all_plan_activated_users = Owner.objects.filter(
+                ownerid__in=all_plan_activated_user_ids
+            ).prefetch_related("user")
+            non_student_count = all_plan_activated_users.exclude(student=True).count()
```
Optional: maybe adding a comment on why we exclude students. Thanks for being so attentive and catching this this early, this is a edge case for suuuure

nora-codecov replied:
hmmm, not really an edge case - check out `activated_user_count` or `can_activate_user` methods on `Owner` and `Account` - it's our built in policy.
I'm following the methods defined on the models to calculate things like `can_activate_user` - but having to write it out instead of using the pre-built methods so that I can do bulk operations 👍

adrian-codecov replied:
Right, what I meant was that this is an easy thing to slip since the majority of our customers aren't students, so I was praising you for being diligent, looking at previous code and learning about our policy

adrian-codecov commented on codecov_auth/admin.py:
```diff
@@ -310,6 +311,66 @@ class AccountAdmin(AdminMixin, admin.ModelAdmin):
+            )
+            userless_owners = []
+            for userless_owner in owners_without_user_objects:
+                new_user = User.objects.create()
```
Is this new User object meant to be empty?

nora-codecov replied:
no! fixed!
ty ty ty

adrian-codecov commented on codecov_auth/admin.py:
```diff
@@ -310,6 +311,66 @@ class AccountAdmin(AdminMixin, admin.ModelAdmin):
+            total = Owner.objects.bulk_update(userless_owners, ["user"])
+            self.message_user(
+                request,
+                f"Created a User for {total} Owners",
```
Would probably be handy to log the owner ids we did this for 👀, up to you

adrian-codecov reviewed (commented):
Left some comments, overall lgtm! Lmk and I can approve

michelletran-codecov reviewed (approved):
Generally LGTM! This is only for adding orgs for now, correct? We would need some different mechanism if we were to remove orgs from accounts...

nora-codecov commented:
updated and added based on @adrian-codecov  and @michelletran-codecov feedback! ty for the comments ❤️ 
- added logs  🤖 
- added logic to remove AccountUsers so now the action is a "true" sync  🔄 
- improved names  💪
